modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/endofyear/EndOfYearSync.kt

```diff
@@ -0,0 +1,118 @@
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.time.LocalDate
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.supervisorScope
+
+@Singleton
+class EndOfYearSync @Inject constructor(
+    private val syncManager: SyncManager,
+    private val historyManager: HistoryManager,
+    private val ratingsManager: RatingsManager,
+    private val episodeManager: EpisodeManager,
+    private val settings: Settings,
+    private val clock: Clock,
+) {
+    private val ratingsSync = CachedAction<Unit, Unit> { syncRatings() }
+    private val thisYearSync = CachedAction<Int, Unit> { syncHistoryForYear(it) }
+    private val lastYearSync = CachedAction<Int, Unit> { syncHistoryForYear(it) }
+
+    suspend fun sync(year: Int = EndOfYearManager.YEAR_TO_SYNC): Boolean {
+        return if (FeatureFlag.isEnabled(Feature.END_OF_YEAR_2024) && syncManager.isLoggedIn()) {
+            supervisorScope {
+                val jobs = listOf(
+                    thisYearSync.run(year, scope = this),
+                    lastYearSync.run(year - 1, scope = this),
+                    ratingsSync.run(scope = this),
+                )
+                runCatching { jobs.awaitAll() }
+                    .map { true }
+                    .getOrElse { error -> if (error !is CancellationException) false else throw error }
+            }
+        } else {
+            false
+        }
+    }
+
+    suspend fun reset() {
+        ratingsSync.reset()
+        thisYearSync.reset()
+        lastYearSync.reset()
+    }
+
+    private suspend fun syncRatings() {
+        val lastSyncTimeString = settings.getLastModified()
+        val lastSyncTime = runCatching { Instant.parse(lastSyncTimeString) }.getOrDefault(Instant.EPOCH)
+        // Check if we synced account in the last day and assume that ratings are accurate if that's the case
+        if (Duration.between(lastSyncTime, clock.instant()).toDays() < 1) {
+            return
+        }
+
+        val ratings = syncManager.getPodcastRatings()?.podcastRatingsList
+        if (ratings == null) {
+            throw RuntimeException("Failed to sync ratings")
+        }
+        val userRatings = ratings.mapNotNull { rating ->
+            val modifiedAt = rating.modifiedAt.toDate() ?: return@mapNotNull null
+            UserPodcastRating(
+                podcastUuid = rating.podcastUuid,
+                rating = rating.podcastRating,
+                modifiedAt = modifiedAt,
+            )
+        }
+        ratingsManager.updateUserRatings(userRatings)
+    }
+
+    private suspend fun syncHistoryForYear(year: Int) {
+        // check for an episode interacted with before this year and assume they have the full listening history if they exist
+        if (anyEpisodeInteractionBeforeYear(year)) {
+            return
+        }
```

Comment:
I remember a case where a user had a listening history for 2022. However, they did not access the app on their secondary Android device between Jan and Nov 2023. They continued to use the app on their primary device during that period. Because of this, they noticed a mismatch in their listening time comparison from last year between different devices.

The issue was only resolved when the app was reinstalled on the secondary device.

I think that removing this condition will fix the problem.

Reply:
Yes, I've been wondering about it as well and if should keep this check when migrating syncing to this process. On the one hand it's quite clever but on the other it fails for scenarios that you mentioned.

I'm wondering if there is any check that we could do to avoid network requests. Maybe I could store the eoy sync date in the preferences and check that. What do you think?

Reply:
> Maybe I could store the eoy sync date in the preferences and check that. What do you think?

That makes sense to me. How long would you wait to allow another network request?

Reply:
I think something in the range of 6 to 12 hours is ok. This would only affect users listening on multiple devices so it shouldn't be that big of a deal to have this accuracy window.

Reply:
I've approved the PR. Feel free to tackle it in another PR.

Reply:
I'll add it here to avoid changing this file in the future PRs.